components/eamxx/src/share/field/field_utils_impl_colred.hpp

```diff
@@ -0,0 +1,123 @@
+
+// Utility to compute the reduction of a field along its column dimension.
+// This is equivalent to einsum('i,i...k->...k', f1, f2); i is the column.
+// The layouts are such that:
+// - The first dimension is for the columns (COL)
+// - There can be only up to 3 dimensions
+
+template <typename ST>
+Field column_reduction(const Field &f1, const Field &f2, const ekat::Comm *co) {
+  using KT          = ekat::KokkosTypes<DefaultDevice>;
+  using RangePolicy = Kokkos::RangePolicy<Field::device_t::execution_space>;
+  using TeamPolicy  = Kokkos::TeamPolicy<Field::device_t::execution_space>;
+  using TeamMember  = typename TeamPolicy::member_type;
+  using ESU         = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+  using namespace ShortFieldTagsNames;
+
+  const auto &l1 = f1.get_header().get_identifier().get_layout();
+
+  EKAT_REQUIRE_MSG(l1.rank() == 1,
+                   "Error! First field f1 must be rank-1.\n"
+                   "The input has rank "
+                       << l1.rank() << ".\n");
+  EKAT_REQUIRE_MSG(l1.tags() == std::vector<FieldTag>({COL}),
+                   "Error! First field f1 must have a column dimension.\n"
+                   "The input f1 layout is "
+                       << l1.tags() << ".\n");
+
+  const auto &n2 = f2.get_header().get_identifier().name();
+  const auto &l2 = f2.get_header().get_identifier().get_layout();
+  const auto &u2 = f2.get_header().get_identifier().get_units();
+  const auto &g2 = f2.get_header().get_identifier().get_grid_name();
+
+  EKAT_REQUIRE_MSG(l2.rank() <= 3,
+                   "Error! Second field f2 must be at most rank-3.\n"
+                   "The input f2 rank is "
+                       << l2.rank() << ".\n");
+  EKAT_REQUIRE_MSG(l2.tags()[0] == COL,
+                   "Error! Second field f2 must have a column dimension.\n"
+                   "The input f2 layout is "
+                       << l2.tags() << ".\n");
+  EKAT_REQUIRE_MSG(
+      l1.dim(0) == l2.dim(0),
+      "Error! The two input fields must have the same dimension along "
+      "which we are taking the reducing the field.\n"
+      "The first field f1 has dimension "
+          << l1.dim(0)
+          << " while "
+             "the second field f2 has dimension "
+          << l2.dim(0) << ".\n");
+
+  auto v1 = f1.get_view<const ST *>();
+
+  FieldIdentifier fo_id(n2 + "_colred", l2.clone().strip_dim(0), u2, g2);
+  Field fo(fo_id);
+  fo.allocate_view();
+  fo.deep_copy(0);
+
+  const int d0 = l2.dim(0);
+
+  switch(l2.rank()) {
+    case 1: {
+      auto v2 = f2.get_view<const ST *>();
+      auto vo = fo.get_view<ST>();
+      Kokkos::parallel_reduce(
+          fo.name(), RangePolicy(0, d0),
+          KOKKOS_LAMBDA(const int i, ST &ls) { ls += v1(i) * v2(i); }, vo);
+    } break;
+    case 2: {
+      auto v2      = f2.get_view<const ST **>();
+      auto vo      = fo.get_view<ST *>();
+      const int d1 = l2.dim(1);
+      auto p       = ESU::get_default_team_policy(d1, d0);
```

Comment:
It seems that the team policies are backward here. `nlev` should be the inner loop size, while the product of the remaining dims should be the outer loop size. Perhas you mixed up some stuff from the global horiz integral diag from the other branch you're working on?

Reply:
I usually review "commits since my last review", so you can go bigger now. Also, while I agree on keeping PRs manageable for the reviewer, I don't want to integrate half-features, so if it needs 300 more lines to be complete, so be it. Also, if the 100s of lines are all in the same 2-3 files, it's much easier to follow than a PR of just 200 lines, but scattered across 10-20 files...